parlai/core/logging_utils.py

```diff
@@ -0,0 +1,132 @@
+                file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
+            self.fileHandler = logging.FileHandler(filename)
+            self.fileHandler.level = (
+                file_level
+            )  # Log to file levels: file_level and above
+            self.fileFormatter = logging.Formatter(file_format)
+            self.fileHandler.setFormatter(self.fileFormatter)
+            super().addHandler(self.fileHandler)
+
+        # Logging to stdout
+        self.streamHandler = logging.StreamHandler(sys.stdout)
+        self.streamHandler.level = (
+            console_level
+        )  # Log to stdout levels: console_level and above
+        if console_format is not None:
+            self.consoleFormatter = logging.Formatter(console_format)
+            self.streamHandler.setFormatter(self.consoleFormatter)
+        super().addHandler(self.streamHandler)
+
+        # To be used with testing_utils.capture_output()
+        self.altStream = None
+
+    def log(self, msg, level=INFO):
+        """ Default Logging function """
+        super().log(level, msg)
+
+    def add_file_handler(self, filename, level=INFO, format=None):
+        """
+        Add a file handler to the logger object
+        Use case: When logging using the logger object instead of instantiating a new ParlaiLogger
+                  this function might  be useful to add a filehandler on the go
+        Only does so if there is no file handler existing
+        """
+        if not hasattr(self, 'fileHandler'):
+            if format is None:
+                file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
+            self.fileHandler = logging.FileHandler(filename)
+            self.fileHandler.level = level  # Log to file levels: level and above
+            self.fileFormatter = logging.Formatter(file_format)
+            self.fileHandler.setFormatter(self.fileFormatter)
+            super().addHandler(self.fileHandler)
+        else:
+            self.info("ParlaiLogger: A filehandler already exists")
+
+    def add_format_prefix(self, prefix):
+        """Include `prefix` in all future logging statements"""
+        # change both handler formatters to add a prefix
+        new_str = prefix + " " + '%(message)s'
+        self.streamHandler.setFormatter(logging.Formatter(new_str))
+        if hasattr(self, 'fileHandler'):
+            prevFileFormat = self.fileFormatter._fmt.split('::')[:-1]
+            prevFileFormat += [' ' + new_str]
+            updatedFileFormat = '::'.join(prevFileFormat)
+            self.fileHandler.setFormatter(logging.Formatter(updatedFileFormat))
+
+    def reset_formatters(self):
+        """Resort back to initial formatting"""
+        if hasattr(self, 'fileHandler'):
+            self.fileHandler.setFormatter(self.fileFormatter)
+        self.streamHandler.setFormatter(logging.Formatter('%(message)s'))
+
+    def mute(self):
+        """Stop logging to stdout"""
+        self.streamHandler.level = float('inf')
```

Comment:
is it okay if level isn't an int? I reckon it is but just wondering

Reply:
Tried it out and it seemed to work so I guess it's okay. Got this solution off of SO.